tensorflow/compiler/mlir/xla/transforms/cycle_detector.cc

```diff
@@ -0,0 +1,335 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/mlir/xla/transforms/cycle_detector.h"
+
+#include <algorithm>
+#include "llvm/ADT/DenseSet.h"
+
+namespace mlir {
+
+namespace {
+
+using NodeSet = llvm::DenseSet<int32_t>;
+using OrderedNodeSet = OrderedSet<int32_t>;
+
+template <typename T>
+struct VecStruct {
+  typedef llvm::SmallVector<T, 4> type;
+};
+template <typename T>
+using Vec = typename VecStruct<T>::type;
+
+struct Node {
+  int32_t rank;    // rank number assigned by Pearce-Kelly algorithm
+  bool visited;  // Temporary marker used by depth-first-search
+  void* data;    // User-supplied data
+  OrderedNodeSet in;   // List of immediate predecessor nodes in graph
+  OrderedNodeSet out;  // List of immediate successor nodes in graph
+};
+
+}  // namespace
+
+
+struct GraphCycles::Rep {
+  Vec<Node*> nodes_;
+  Vec<int32_t> free_nodes_;  // Indices for unused entries in nodes_
+
+  // Temporary state.
+  Vec<int32_t> deltaf_;  // Results of forward DFS
+  Vec<int32_t> deltab_;  // Results of backward DFS
+  Vec<int32_t> list_;    // All nodes to reprocess
+  Vec<int32_t> merged_;  // Rank values to assign to list_ entries
+  Vec<int32_t> stack_;   // Emulates recursion stack when doing depth first search
+};
+
+GraphCycles::GraphCycles() : rep_(new Rep) {}
+
+GraphCycles::~GraphCycles() {
+  for (Vec<Node*>::size_type i = 0; i < rep_->nodes_.size(); i++) {
+    delete rep_->nodes_[i];
+  }
+  delete rep_;
+}
+
+int32_t GraphCycles::NewNode() {
+  if (rep_->free_nodes_.empty()) {
+    Node* n = new Node;
```

Comment:
This is a light version of GraphCycles implemented in `tensorflow/compiler/jit/graphcycles/graphcycles.h`

We re-implement it here because we do not want to rely on tensorflow data structures as suggested by @joker-eph . This class is relatively independent of the fusion pass itself and may have some other users.  The original design also supports to remove node after allocated it, thus the node ids may not continuous if we allocate a sequence of nodes. This implies that we need to return a vector of ids if we support interface like `NewNodes(int num_nodes_to_allocate)`, which may be wasteful for the allocation of a new vector.

One possible solution is that we re-design class `GraphCycles` and totally specialize it for the fusion pass. This may not really be helpful specially we already have an implementation ^^